lib/sshutils/x11/forward.go

```diff
@@ -0,0 +1,229 @@
+package x11
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/gravitational/teleport/lib/sshutils"
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	// DefaultDisplayOffset is the default display offset when
+	// searchinf for an X11 Server reverse tunnel port.
+	DefaultDisplayOffset = 10
+	// DisplayEnv is an environment variable used to determine what
+	// local display should be connected to during x11 forwarding.
+	DisplayEnv = "DISPLAY"
+
+	// x11BasePort is the base port used for xserver tcp addresses.
+	// e.g. DISPLAY=localhost:10 -> net.Dial("tcp", "localhost:6010")
+	// Used by some XServer clients, such as openSSH and MobaXTerm.
+	x11BasePort = 6000
+	// x11MaxDisplays is the number of displays which the
+	// server will support concurrent x11 forwarding for.
+	x11MaxDisplays = 1000
+	// x11SocketDir is the name of the directory where x11 unix sockets are kept.
+	x11SocketDir = ".X11-unix"
+)
+
+// ServerConfig is a server configuration for x11 forwarding
+type ServerConfig struct {
+	// Enabled controls whether x11 forwarding requests can be granted by the server.
+	Enabled bool `yaml:"enabled"`
+	// DisplayOffset tells the server what x11 display number to start from.
+	DisplayOffset int `yaml:"display_offset,omitempty"`
+}
+
+// ForwardRequestPayload according to http://www.ietf.org/rfc/rfc4254.txt
+type ForwardRequestPayload struct {
+	// SingleConnection determines whether any connections will be forwarded
+	// after the first connection, or after the session is closed. In OpenSSH
+	// and Teleport SSH clients, SingleConnection is always set to false.
+	SingleConnection bool
+	// AuthProtocol is the name of the X11 authentication protocol being used.
+	AuthProtocol string
+	// AuthCookie is a hexadecimal encoded X11 authentication cookie. This should
+	// be a fake, random cookie, which will be checked and replaced by the real
+	// cookie once the connection request is received.
+	AuthCookie string
+	// ScreenNumber determines which screen will be.
+	ScreenNumber uint32
+}
+
+// X11ChannelRequestPayload according to http://www.ietf.org/rfc/rfc4254.txt
+type X11ChannelRequestPayload struct {
+	// OriginatorAddress is the address of the server requesting an x11 channel
+	OriginatorAddress string
+	// OriginatorPort is the port of the server requesting an x11 channel
+	OriginatorPort uint32
+}
+
+// RequestX11Forwarding sends an "x11-req" to the server to set up x11 forwarding for the given session.
+// authProto and authCookie are required to set up authentication with the Server. screenNumber is used
+// by the server to determine which screen should be connected to for x11 forwarding. singleConnection is
+// an optional argument to request x11 forwarding for a single connection.
+func RequestX11Forwarding(sess *ssh.Session, display, authProto, authCookie string) error {
```

Comment:
This will always be called with an `x11` prefix, so how about we name it `x11.RequestForwarding()`?

(And similar comment for `ServeX11ChannelRequests`)